libcnb/src/tracing.rs

```diff
@@ -0,0 +1,136 @@
+use libcnb_data::buildpack::Buildpack;
+use opentelemetry::{
+    global,
+    trace::{Span as SpanTrait, Status, Tracer, TracerProvider as TracerProviderTrait},
+    KeyValue,
+};
+use opentelemetry_sdk::{
+    trace::{Config, Span, TracerProvider},
+    Resource,
+};
+use std::path::Path;
+
+const TELEMETRY_EXPORT_ROOT: &str = "/tmp/cnb-telemetry";
+
+pub(crate) struct BuildpackTrace {
+    provider: TracerProvider,
+    span: Span,
+}
+
+pub(crate) fn start_trace(buildpack: &Buildpack, phase_name: &'static str) -> BuildpackTrace {
+    let trace_name = format!("{}-{phase_name}", buildpack.id.replace(['/', '.'], "_"));
+    let tracing_file_path = Path::new(TELEMETRY_EXPORT_ROOT).join(format!("{trace_name}.jsonl"));
+
+    // Ensure tracing file path parent exists by creating it.
+    if let Some(parent_dir) = tracing_file_path.parent() {
+        let _ = std::fs::create_dir_all(parent_dir);
+    }
+    let exporter = match std::fs::File::options()
+        .create(true)
+        .append(true)
+        .open(&tracing_file_path)
+    {
+        Ok(file) => opentelemetry_stdout::SpanExporter::builder()
+            .with_writer(file)
+            .build(),
+        Err(_) => opentelemetry_stdout::SpanExporter::builder()
+            .with_writer(std::io::sink())
+            .build(),
+    };
+
+    let lib_name = option_env!("CARGO_PKG_NAME").unwrap_or("libcnb");
+    let lib_version = option_env!("CARGO_PKG_VERSION");
+
+    let provider = TracerProvider::builder()
+        .with_simple_exporter(exporter)
+        .with_config(Config::default().with_resource(Resource::new(vec![
+            KeyValue::new("service.name", lib_name),
+            KeyValue::new("service.version", lib_version.unwrap_or_default()),
+        ])))
+        .build();
+
+    global::set_tracer_provider(provider.clone());
+
+    let tracer = provider.versioned_tracer(lib_name, lib_version, None as Option<&str>, None);
+
+    let mut span = tracer.start(trace_name);
+    span.set_attributes(vec![
+        KeyValue::new("buildpack_id", buildpack.id.to_string().clone()),
+        KeyValue::new("buildpack_name", buildpack.name.clone().unwrap_or_default()),
+        KeyValue::new("buildpack_version", buildpack.version.to_string()),
+        KeyValue::new(
+            "buildpack_homepage",
+            buildpack.homepage.clone().unwrap_or_default(),
+        ),
```

Comment:
I added these mostly because they were available, and I've been told in the past that it's ok for wide events to be really wide (assuming you aren't generating keys). I'm also unsure how useful they'll be. Maybe lets try it and we can remove later?